### PR TITLE
PUBDEV-7232 - Remove DeepWater Estimator Function

### PR DESCRIPTION
**Pull request by Pscheidl**
https://0xdata.atlassian.net/browse/PUBDEV-7232

- Remove it from "everywhere"
- Leave it in GenModel (backwards compatibility)
- Remove jupyter notebooks containining only DeepWater models. Remove DeepWater from those notebooks which contains anything else than DeepWater.
- The README in deeplearning examples seems to be a little bit outdated. Some of the examples containing DW were not even listed.
- Should we leave the Py/R API in place and only throw informative errors ? If we delete it once and for all, it should be pretty obvious it has been removed.

### CODE REVIEW

**Comment by sebhrusen** [comment]
quick question: is it reasonable to remove a feature in a `release` branch? should we not do this in `master` for next release instead?

**Comment by Pscheidl** [comment]
Quick answer: I'm not 100 percent sure. I think users should be able to switch between fix releases without features disappearing.

**Comment by michalkurka** [comment]
> Leave it in GenModel (backwards compatibility)

Can we please remove it from genmodel as well? Or maybe move it from regular genmodel to h2o-genmodel-extension/deepwater? And don't bundle this module by default to the genmodel? If anyone really needs it they will have the deepwater-genmodel jar.

**Comment by Pscheidl** [comment]
That sounds reasonably well @michalkurka. I'll make it so.

**Comment by Pscheidl** [comment]
> > Leave it in GenModel (backwards compatibility)
> 
> Can we please remove it from genmodel as well? Or maybe move it from regular genmodel to h2o-genmodel-extension/deepwater? And don't bundle this module by default to the genmodel? If anyone really needs it they will have the deepwater-genmodel jar.

Done. Please see the changes in the following commit: https://github.com/h2oai/h2o-3/pull/4393/commits/6d911767ea6e50babafa7a7a7e73ab29a091fc03

Also, please note there was an addition to GenModel - a new method named `getRawDataConverter`. Please see if that's ok for you and if not, we might be able to do it in some other way. With extensions, I can see no other option than a strategy pattern.

**Comment by honzasterba** [comment]
Definitely remove the py/r api, thats error enough.

File: settings.gradle [hunk 1]
```diff
@@ -39,6 +39,7 @@ include 'h2o-security'
 include 'h2o-logger'
 include 'h2o-genmodel-extensions:jgrapht'
 include 'h2o-k8s'
+include 'h2o-genmodel-extensions:deepwater'
```
Comment (michalkurka) [inline settings.gradle]: this should imho follow the xgboost naming convention, h2o-genmodel-ext-deepwater

File: settings.gradle [hunk 2]
```diff
@@ -39,6 +39,7 @@ include 'h2o-security'
 include 'h2o-logger'
 include 'h2o-genmodel-extensions:jgrapht'
```
Comment (michalkurka) [inline settings.gradle]: also not following convention, should be imho fixed as well @koniecsveta

File: settings.gradle (hunk 1 above)
Comment (Pscheidl) [reply settings.gradle]: Right now, it seems like me and @koniecsveta set the conventions :rofl: 

Of course, I'll rename it.

**Comment by Pscheidl** [comment]
Waiting for tests, @sebhrusen please review if you have a minut or two.

**Comment by Pscheidl** [comment]
`EasyPredictModelWrapperTest` failure. Investigating.

**Comment by Pscheidl** [comment]
```
java.lang.NullPointerException
	at hex.genmodel.easy.EasyPredictModelWrapper.fillRawData(EasyPredictModelWrapper.java:850)
	at hex.genmodel.easy.EasyPredictModelWrapper.predict(EasyPredictModelWrapper.java:855)
	at hex.genmodel.easy.EasyPredictModelWrapper.preamble(EasyPredictModelWrapper.java:838)
	at hex.genmodel.easy.EasyPredictModelWrapper.predictRegression(EasyPredictModelWrapper.java:766)
	at hex.genmodel.easy.EasyPredictModelWrapper.predictRegression(EasyPredictModelWrapper.java:754)
	at hex.genmodel.easy.EasyPredictModelWrapperTest.testScoreOffsetRegression(EasyPredictModelWrapperTest.java:254)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:33)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:230)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:58)
```

rowDataConverter is null. This is related to these changes.

**Comment by Pscheidl** [comment]
Fixed by telling Mockito to call the real method - it was not called and null had been returned.

**Comment by Pscheidl** [comment]
I need this PR: https://github.com/h2oai/mojoland/pull/12

to solve the MOJO compatibility check test failure. Please see and review.

**Comment by Pscheidl** [comment]
Had another idea how to solve this via Service Loader. Waiting for tests.

File: h2o-genmodel/src/main/java/hex/genmodel/CommonRowtoRawDataConverterProvider.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
+package hex.genmodel;
+
+import hex.genmodel.easy.CategoricalEncoder;
+import hex.genmodel.easy.EasyPredictModelWrapper;
+import hex.genmodel.easy.RowToRawDataConverter;
+
+import java.util.Map;
+
+public class CommonRowtoRawDataConverterProvider implements RowtoRawDataConverterProvider {
```
Comment (honzasterba) [inline h2o-genmodel/src/main/java/hex/genmodel/CommonRowtoRawDataConverterProvider.java]: I would prefer DefaultRowtoRawDataConverterProvider over Common*
other than that I think this is great

File: h2o-genmodel/src/main/java/hex/genmodel/easy/EasyPredictModelWrapper.java [hunk 4]
```diff
@@ -301,12 +301,29 @@ public EasyPredictModelWrapper(Config config) {
...
     }
 
-    CategoricalEncoding categoricalEncoding = config.getUseExternalEncoding() ? 
+    CategoricalEncoding categoricalEncoding = config.getUseExternalEncoding() ?
             CategoricalEncoding.AUTO : m.getCategoricalEncoding();
     Map<String, Integer> columnMapping = categoricalEncoding.createColumnMapping(m);
     Map<Integer, CategoricalEncoder> domainMap = categoricalEncoding.createCategoricalEncoders(m, columnMapping);
 
-    rowDataConverter = RowDataConverterFactory.makeConverter(m, columnMapping, domainMap, errorConsumer, config);
+    final RowtoRawDataConverterProvider rowtoRawDataConverterProvider = getRawDataConverterProviderForGenmodel(m);
```
Comment (michalkurka) [inline h2o-genmodel/src/main/java/hex/genmodel/easy/EasyPredictModelWrapper.java]: Can we have a `makeConverterFactory` on MojoModel, `makeConvert` method can then be defined on the mojo level with appropriate non-public visibility.

I would go for a simpler solution here.

**Comment by Pscheidl** [comment]
I implemented one of Michal's proposals. Not as decoupled, but definitely simpler. I agree this solutions feels right. WDYT ? Please re-review gentlemen.

File: h2o-genmodel-extensions/deepwater/build.gradle [hunk 5]
```diff
@@ -0,0 +1,13 @@
+description = "H2O GenModel Deepwater Extension"
+
+sourceCompatibility = 1.7
+targetCompatibility = 1.7
+
+dependencies {
+    compile project(":h2o-genmodel")
```
Comment (honzasterba) [inline h2o-genmodel-extensions/deepwater/build.gradle]: indent

Comment (Pscheidl) [reply h2o-genmodel-extensions/deepwater/build.gradle]: Thank you. Fixed. Pushing soon.

**Pull request merged**